### ISSUE TITLE
TS2869 not fired on `stubTrue() ?? false`

### ISSUE THREAD

**Issue by fregante** [microsoft/TypeScript#62119]
### 🔎 Search Terms

TS 2869
Right operand of ?? is unreachable because the left operand is never nullish

### 🕗 Version & Regression Information

TS2869 was introduced in v5.6 and all version until 5.9-nightly are affected.

### ⏯ Playground Link

https://www.typescriptlang.org/play/?ts=5.9.0-dev.20250724#code/C4JwrgpgBA-DUDMCGAbAzhA3AWAFAAp8BKKAXgD4pRIS5FUMcDizLqIiW7l0I8g

### 💻 Code

```ts
true ?? false;             // detected
(() => true) ?? false;     // detected
(() => true)() ?? false    // not detected
```

### 🙁 Actual behavior

`(() => true)() ?? false` does not report any errors

### 🙂 Expected behavior

```ts
(() => true)() ?? false
~~~~~~~~~~~~~~
             ^ Right operand of ?? is unreachable because the left operand is never nullish
```

**Comment by RyanCavanaugh** [comment]
This check is intentionally only syntactic, and a function call syntactically may return a nullish value. Trying to "fix" this would just introduce inconsistencies or false positives.

**Comment by fregante** [comment]
> a function call syntactically may return a nullish value


In what scenario a function typed as `() => boolean` returns a nullish value?

**Comment by RyanCavanaugh** [comment]
> In what scenario a function **typed** as

This check is **syntactic**; the types are not part of the analysis

**Comment by fregante** [comment]
I thought that typescript was checking my types. 

I get that it was intentional, but it definitely feels like it could very well do better.

**Comment by RyanCavanaugh** [comment]
See #59217 - nullishness can creep in through JS calls, array access with noUncheckedIndexedAccess off, exterior mutation, etc

**Comment by fregante** [comment]
> nullishness can creep in through JS calls

But that's the case for all checks. Should Typescript be considered unsafe even in strict mode? Why is `stubTrue() ?? false` ignored but `stubTrue() === undefined` is considered an error?

**Comment by fregante** [comment]
Good thing there's TS-ESLint

https://typescript-eslint.io/play/#ts=5.8.3&fileType=.ts&code=C4JwrgpgBA-DUDMCGAbAzhA3AKABS4EooBeAPilEiLkVQx3yLIvAgMNnmXQiA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK%2BSpPRRE0aB2iRwYAL4gtQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

Not sure why I did not have this rule in my repo already

<img width="595" height="208" alt="Image" src="https://github.com/user-attachments/assets/b6c62a64-2f7b-443d-81af-b83897504445" />

**Comment by fregante** [comment]
Tickets related to the problem you're describing:

- https://github.com/microsoft/TypeScript/issues/36393
- https://github.com/typescript-eslint/typescript-eslint/issues/1798
